### PR TITLE
Start a topology chapter with a filters section

### PR DESCRIPTION
**Pull request by PatrickMassot**

### CODE REVIEW

File: lean_source/07_Topology/source_01_Filters.lean [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+The pull-back operation is also compatible with composition, but contravariant.
+TEXT. -/
+
+-- QUOTE:
+#check @comap_comap
+-- QUOTE.
+
+/- TEXT:
+Let's now move up to the plane ``ℝ × ℝ`` and try to understand how neighborhoods of some point
+``(x₀, y₀)`` are related to ``𝓝 x₀`` and ``𝓝 y₀``. There is product operation
```
Comment (Julian) [inline lean_source/07_Topology/source_01_Filters.lean:226]:
```suggestion
``(x₀, y₀)`` are related to ``𝓝 x₀`` and ``𝓝 y₀``. There is a product operation 
```

File: lean_source/07_Topology/source_01_Filters.lean [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+  unfold tendsto filter.prod,
+  rw [le_inf_iff, ← map_le_iff_le_comap, map_map, ← map_le_iff_le_comap, map_map]
+end -/
+
+/- TEXT:
+The ordered type ``filter X`` is actually a *complete* lattice: there is a bottom element and a top element and
+any set of filters on ``X`` have an ``Inf`` and a ``Sup``.
+
+Note that in the definition of filters, given the second property
+(if ``U`` belong to ``F`` then anything larger also belongs), the first property
```
Comment (Julian) [inline lean_source/07_Topology/source_01_Filters.lean:279]:
```suggestion
(if ``U`` belongs to ``F`` then anything larger also belongs), the first property 
```

File: lean_source/07_Topology/source_01_Filters.lean [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+definition does not prohibit ``∅ ∈ F``. If ``∅ ∈ F`` then
+``∀ U : set X, U ∈ F``, hence ``F`` is a
+rather trivial filter. This trivial filter is precisely the bottom element of the complete
+lattice ``filter X``.
+This contrasts with the definition of filters in
+Bourbaki who doesn't allow filters containing the empty set. 
+Because of this choice, we need an
+extra assumption of non-triviality in some lemmas, but gain
+nicer global formal properties for the theory. We've seen already that it gives us a
+bottom element. It also allowed us to define ``principal : set X → filter X`` without
```
Comment (Julian) [inline lean_source/07_Topology/source_01_Filters.lean:292]:
```suggestion
bottom element. It also allows us to define ``principal : set X → filter X`` without
```

File: lean_source/07_Topology/source_01_Filters.lean [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+
+-- QUOTE.
+
+/- TEXT:
+
+Readers who know about measure theory will note that the filter ``μ.ae`` of sets whose complement has measure zero
+(aka "the set of almost every point") is not very useful as the source or target of ``tendsto`` but very often
+used with ``eventually``, when saying a property holds for almost every point.
+
+There is dual version of ``∀ᶠ x in F, P x`` which is much less useful, but still exists: ``∃ᶠ x in F, P x`` means
```
Comment (Julian) [inline lean_source/07_Topology/source_01_Filters.lean:440]:
```suggestion
There is a dual version of ``∀ᶠ x in F, P x`` which is much less useful, but still exists: ``∃ᶠ x in F, P x`` means
```

File: lean_source/07_Topology/source_01_Filters.lean [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+``{x | ¬P x} ∉ f``. For instance ``∃ᶠ n in at_top, P n`` means there are arbitrarily large ``n`` such that ``P n``.
+This notation stands for ``filter.frequently``.
+
+
+As a more sophisticated example, the statement: "if a sequence ``u`` converges to
+some ``x`` and ``u n`` belongs to a set ``M`` for ``n`` large enough then ``x`` is in the closure of
+``M``" is formalized as: ``tendsto u at_top (𝓝 x) → (∀ᶠ n in at_top, u n ∈ M) → x ∈ closure M``,
+which is a special case of ``mem_closure_of_tendsto`` from the topology library.
+See if you can prove it using the quoted lemmas, knowing that ``cluster_pt x F`` means ``(𝓝 x ⊓ F).ne_bot``.
```
Comment (Julian) [inline lean_source/07_Topology/source_01_Filters.lean:449]:
```suggestion
See if you can prove it using the quoted lemmas, keeping in mind that ``cluster_pt x F`` means ``(𝓝 x ⊓ F).ne_bot``.
```

**Review by Julian — COMMENTED** [review]
Some basic copyediting as I tried to read some of this, in case it's helpful.

File: lean_source/07_Topology/source_01_Filters.lean [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+
+Filters
+-------
+
+Filters on a type ``X`` are collections of sets of ``X`` satisfying three conditions. They are mostly used to
+abstract two related kinds of ideas:
+
+* *limits*, including including all kinds of limites discussed above, in particular finite or infinite limits of sequences, finite or infinite limits of functions at a point or at infinity, etc...
+
+* *things happening eventually*, including things happening for large enough ``n : ℕ``, or near enough a point ``x``, or for close enough pairs of points, or things happening almost everywhere in the sense of measure theory. Dually, filters can also express the idea of *things happening often*: for arbitrarily large ``n``, or at a point in any neighborhood of given a point etc...
```
[the remark this answers is not in the document]
Comment (slel) [reply lean_source/07_Topology/source_01_Filters.lean:19]: I think it's "near a point" or "close to a point", not "near to a point".

File: lean_source/07_Topology/source_01_Filters.lean [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+
+The examples of filters appearing in these descriptions will be described in details later, but we can already name them:
+
+* ``(at_top : filter ℕ)`` : made of sets of ``ℕ`` containing ``{n | n ≥ N}`` for some ``N``
+* ``𝓝 x`` : made of neighborhoods of ``x`` in a topological space
+* ``𝓤 X`` : made of entourages of a uniform space (those space are generalizations of metric spaces and topological groups)
```
[the remark this answers is not in the document]
Comment (slel) [reply lean_source/07_Topology/source_01_Filters.lean:25]: Or: "uniform spaces generalize metric spaces and topological spaces".

File: lean_source/07_Topology/source_01_Filters.lean (hunk 3 above)
Comment (PatrickMassot) [reply lean_source/07_Topology/source_01_Filters.lean:292]: I don't understand why you suggest this changed. The definition of `principal` indeed appeared before in that file.

File: lean_source/07_Topology/source_01_Filters.lean (hunk 5 above)
Comment (PatrickMassot) [reply lean_source/07_Topology/source_01_Filters.lean:449]: It doesn't seem fair to write "keeping in mind" when this has never been explained.

**Comment by PatrickMassot** [comment]
Thanks you very much Julian and Samuel! I'm sorry I left so many typos. This certainly comes from me being very bad at proofreading anything, but also I opened this PR to gather feedback from my coauthors about the global direction of this section, it wasn't meant to be polished.

File: lean_source/07_Topology/source_01_Filters.lean (hunk 3 above)
Comment (Julian) [reply lean_source/07_Topology/source_01_Filters.lean:292]: It's a very minor point, I think stylistically present tense tends to sound better than past tense even if it was indeed earlier in the file, but it's yeah, very minor, leave it if you like it, it's not grammatically incorrect.

**Comment by Julian** [comment]
Patrick happy to help in a small way, and no worries on the polish, as I say I was just myself reading it and thought I may as well leave the comments. Overall I found the content good too!

**Pull request merged**